apps/frontend/src/app/PageCharacter/CharacterDisplay/Tabs/TabTheorycraft/optimizeTc.ts

```diff
@@ -57,92 +78,213 @@ export function optimizeTc(
+    structuredClone(bufferRolls)
+  const existingSubs = objMap(charTC.artifact.substats.stats, (v, k) =>
+    toDecimal(v, k)
   )
-  const subsArr = [...subs]
-  let distributed = distributedSubstats
-  const maxSubstats = objMap(rawMaxSubstats, (v, k) => {
-    return (
-      v -
-      clampLow(
-        Math.ceil(substats[k] / getSubstatValue(k, rarity, substatsType)),
-        0
+  const mainStatsCount = getMainStatsCount(slots)
+  const minSubLines = getMinSubLines(slots)
+
+  const alreadyFeasible =
+    getMinOtherRolls(
+      Object.entries(existingRolls),
+      mainStatsCount,
+      minSubLines
+    ) <= 0
+
+  callback({
+    resultType: 'total',
+    total: countPerms(
+      distributedSubstats,
+      [...scalesWithSub, 'other'].map((k) =>
+        k === 'other' ? distributedSubstats : maxSubsAssignable[k]
       )
-    )
+    ),
   })
-  const assignableMaxTot = subsArr.reduce((a, x) => a + maxSubstats[x], 0)
-  if (assignableMaxTot <= distributedSubstats) {
-    distributed = assignableMaxTot
-    maxBuffer = Object.fromEntries(
-      subsArr.map((x) => [x, substatValue(x, maxSubstats[x])])
-    )
-    if (shouldShowDevComponents)
-      console.log({ maxBuffer, subsArr, maxSubstats, distributed })
-  } else {
-    let max = -Infinity
-    const buffer = Object.fromEntries([...subs].map((x) => [x, 0]))
-    const existingSubs = objMap(
-      charTC.artifact.substats.stats,
-      (v, k) => v / comp(k)
-    )
-    const permute = (toAssign: number, [x, ...xs]: string[]) => {
-      if (xs.length === 0) {
-        if (toAssign > maxSubstats[x]) return
-        buffer[x] = substatValue(x, toAssign)
-        const [result] = compute([
-          { values: existingSubs },
-          { values: buffer },
-        ] as const)
-        if (result > max) {
-          max = result
-          maxBuffer = structuredClone(buffer)
+  let tested = 0
+  let failed = 0
+  let skipped = 0
+  const constraints = Object.entries(minTotal).map(([k, v]) => toDecimal(v, k))
+  const permute = (toAssign: number, [x, ...xs]: string[]) => {
+    if (xs.length === 0) {
+      if (toAssign > maxSubsAssignable[x]) return
+      tested++
+      if (!(tested % 100_000))
+        callback({
+          resultType: 'count',
+          tested,
+          failed,
+          skipped,
+        })
+      if (x !== 'other') buffer[x] = substatValue(x, toAssign)
+      bufferRolls[x] = toAssign
+      if (!alreadyFeasible) {
+        //check for distributed feasibility
+        const allRolls = allSubstatKeys.map((k) => [
+          k,
+          (existingRolls[k] ?? 0) + (bufferRolls[k] ?? 0),
+        ]) as Array<[SubstatKey, number]>
+        const minOtherRolls = getMinOtherRolls(
+          allRolls,
+          mainStatsCount,
+          minSubLines
+        )
+        // not feasible
+        if ((bufferRolls.other ?? 0) < minOtherRolls) {
+          skipped++
+          return
         }
+      }
+      const results = compute([
```

Comment:
same as above, since `existingSubs` is static across all perms, can we precompile this data?
Also, found a weird case, where `existingSubs` seems to be doubled for no reason?
<img width="341" alt="image" src="https://github.com/frzyc/genshin-optimizer/assets/1754901/7f1858f0-a8fa-429c-9bed-3528993c02d7">

Reply:
If there are common values, you can set them in `initial` parameter of `precompute`. On the bug, do you have the (JSON) optimization nodes?

Reply:
Here is the print out of the `nodes`
```
[{"operation":"mul","operands":[{"operation":"const","operands":[],"type":"number","value":1.5130582272},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":1.15},{"operation":"mul","operands":[{"operation":"const","operands":[],"type":"number","value":2.7777777777777777},{"operation":"sum_frac","operands":[{"operation":"add","operands":[{"operation":"read","operands":[],"path":["dyn","eleMas"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","eleMas"],"accu":"add","type":"number"}]},{"operation":"const","operands":[],"type":"number","value":1400}]}]}]},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":1},{"operation":"mul","operands":[{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":2.16753600333786},{"operation":"read","operands":[],"path":["dyn","critDMG_"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","critDMG_"],"accu":"add","type":"number"}]},{"operation":"max","operands":[{"operation":"const","operands":[],"type":"number","value":0},{"operation":"min","operands":[{"operation":"const","operands":[],"type":"number","value":1},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":0.05},{"operation":"read","operands":[],"path":["dyn","critRate_"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","critRate_"],"accu":"add","type":"number"}]}]}]}]}]},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":311},{"operation":"read","operands":[],"path":["dyn","atk"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","atk"],"accu":"add","type":"number"},{"operation":"mul","operands":[{"operation":"const","operands":[],"type":"number","value":713.7632463813659},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":1.35},{"operation":"read","operands":[],"path":["dyn","atk_"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","atk_"],"accu":"add","type":"number"}]}]},{"operation":"mul","operands":[{"operation":"const","operands":[],"type":"number","value":0.008},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":4780},{"operation":"read","operands":[],"path":["dyn","hp"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","hp"],"accu":"add","type":"number"},{"operation":"mul","operands":[{"operation":"const","operands":[],"type":"number","value":15443.342365844726},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":1.916},{"operation":"read","operands":[],"path":["dyn","hp_"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","hp_"],"accu":"add","type":"number"}]}]}]}]},{"operation":"mul","operands":[{"operation":"const","operands":[],"type":"number","value":0.01},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":4780},{"operation":"read","operands":[],"path":["dyn","hp"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","hp"],"accu":"add","type":"number"},{"operation":"mul","operands":[{"operation":"const","operands":[],"type":"number","value":15443.342365844726},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":1.916},{"operation":"read","operands":[],"path":["dyn","hp_"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","hp_"],"accu":"add","type":"number"}]}]}]}]},{"operation":"min","operands":[{"operation":"const","operands":[],"type":"number","value":2855.0529855254636},{"operation":"mul","operands":[{"operation":"const","operands":[],"type":"number","value":0.05957},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":4780},{"operation":"read","operands":[],"path":["dyn","hp"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","hp"],"accu":"add","type":"number"},{"operation":"mul","operands":[{"operation":"const","operands":[],"type":"number","value":15443.342365844726},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":1.916},{"operation":"read","operands":[],"path":["dyn","hp_"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","hp_"],"accu":"add","type":"number"}]}]}]}]}]}]}]},{"operation":"add","operands":[{"operation":"const","operands":[],"type":"number","value":1.49545},{"operation":"read","operands":[],"path":["dyn","enerRech_"],"accu":"add","type":"number"},{"operation":"read","operands":[],"path":["dyn","enerRech_"],"accu":"add","type":"number"}]}]
```

Reply:
Yea, that formula is `1.49545 + <enerRech_> + <enerRech_>`. Do you have `nodes` before `optimize` call on (I assume) line 88?